### PR TITLE
8264744: (fs) Use file cloning in Linux version of Files::copy method

### PR DESCRIPTION
**Pull request by bplb**
Add file cloning to `java.nio.channels.FileChannel::transferTo` and `java.nio.file.Files.copy(Path,Path)`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8264744](https://bugs.openjdk.org/browse/JDK-8264744): (fs) Use file cloning in Linux version of Files::copy method

### Reviewers
 * @mkarg (no known github.com user name / role) ⚠️ Review applies to [d2a862c4](https://git.openjdk.org/jdk/pull/9486/files/d2a862c4404835107a2a5fcd1f75e834548343b9)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9486/head:pull/9486` \
`$ git checkout pull/9486`

Update a local copy of the PR: \
`$ git checkout pull/9486` \
`$ git pull https://git.openjdk.org/jdk pull/9486/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9486`

View PR using the GUI difftool: \
`$ git pr show -t 9486`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9486.diff">https://git.openjdk.org/jdk/pull/9486.diff</a>

</details>

### CODE REVIEW

**Comment by bplb** [comment]
On Linux, the `copy_file_range(2)` system call already includes file cloning if available, so there is no need to add an explicit call to `ioctl(2)` with request code `FICLONE` or `FICLONERANGE` unless `copy_file_range()` is unavailable. Therefore the Linux `ioctl_ficlone` branch in `cloneFile0()` is conditional, and an invocation of `copy_file_range()` is prepended in `directCopy0()` on Linux. Note that as `FileChannel::transferFrom` was already calling `copy_file_range()`, it was already using cloning on Linux when possible.

On macOS, the available cloning system calls are `clonefile(2)` and `copyfile(3)` with for the latter `COPYFILE_CLONE` set in `flags`. There is no cloning call which accepts file descriptors as the target cannot already exist when the system call is invoked. There is also no call on macOS to clone a range of a file, hence there is no cloning added for `transferFrom()` nor `transferTo()` on macOS.

The througput of `Files::copy` is greatly increased by using cloning, at the expense of course of the first subsequent write to the target path being slower due to copy-on-write (CoW).
```
                            throughput (ops/s)
                APFS (macOS)         BTRFS (Linux)         EXT4 (Linux)
count         before   after       before   after        before   after
10240        1578.971 2902.128   24331.624 31074.073   27722.815 27574.891
51200        1849.027 2917.996   19295.797 31313.13    18590.463 18650.098
102400       1757.527 2899.018   15874.133 31243.225   13242.939 13235.246
512000       1195.215 2818.547    5856.964 30972.607    3998.696  3903.092
1048568       829.433 2169.738    3283.841 31097.633    2056.693  2014.939
10485760      126.736 2336.032     236.58  30754.849    154.103    153.356
104857600      15.014 2361.208      22.352 31123.14      14.864     14.746
1073741824      2.484 2335.799       1.549 30682.197      1.408      1.41
```
The throughput within a CoW file system (APFS and BTRFS) is much higher, especially for large files, and that within a non-CoW file system (EXT4) is effectively unchanged.

As APFS supports CoW, existing tests should cover these changes. For Linux, a dynamic search for file systems supporting CoW is added in the `CopyAndMove` test.

It is unclear whether the Linux branch in `cloneFile0` should attempt to use `ioctl_ficlone` if `copy_file_range` is undefined, as currently proposed, or simply always return `IOS_UNAVAILABLE`. It is likewise unclear whether `ioctl_ficlonerange` should be used in `transferFrom0` and `transferTo0` if `copy_file_range` is undefined (it currently is not).

In the initial version of this patch, cloning is enabled by default. In the issue description the possibility of adding a (JDK-specific?) `CopyOption` or system property to disable cloning was mentioned. There is also the opposite approach of cloning being disabled by default, with a `CopyOption` or system property required to enable it.

**Comment by AlanBateman** [comment]
This is good work but it is putting a lot of platform specific code in UnixCopyFile. I think we'll need we'll need to do some up front refactoring to support this so that the copy_file_range and clonefile implementation are Linux and macOS specific source files. This should allow more of the code to be written in Java and will allow most of the native code to go away.

**Comment by AlanBateman** [comment]
This version looks much better but I think it will need another round of two to get the code into the right place. As it stands, the change mean that some of the macOS specific file copy code is in UnixCopyFile and the reset is in BsdNativeDispatcher. 

For macOS I suspect we'll end up with BsdCopyFile extending UnixCopyFile add adding support for fcopyfile and clonefile. Similarly LinuxCopyFile will extend UnixCopyFile and add support for copy_file_range.

A few specific issues on this version:
- LinuxFileSystemProvider.clone leaks a file descriptor if creating the target file fails
- LinuxFileSystemProvider.clone checks if the file exists but that check doesn't make sense here
- LinuxNativeDispatcher.ioctl_ficlone uses Blocker but I would have expected that in the caller, not here
- BsdFileSystemProvider.clone uses noFollowLinks. I think it would be clearer everywhere to invert that to followLinks. It may be better to map it to the flag at the Java level too.
- I wonder if FICLONE is available at build time to avoid the hardcoded value (main concern is whether it has the same value on all Linux ports)

File: test/jdk/java/nio/file/Files/CopyAndMove.java [hunk 1]
```diff
@@ -40,14 +40,64 @@
...
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
 import jdk.test.lib.Platform;
 import jdk.test.lib.RandomFactory;
 
 public class CopyAndMove {
+    // File system types that support copy-on-write (CoW)
+    // for testing cloning / reflink on Linux
+    private static final Set<String> COW_TYPES =
+        Set.of("bcachefs", "btrfs", "ocfs2", "xfs");
```
Comment (mkarg) [inline test/jdk/java/nio/file/Files/CopyAndMove.java:52]: Shouldn't we also test on non-CoW-FS to be sure that the recent changes do not have a negative impact *there*?

**Review by mkarg — APPROVED** [review]
I really appreciate all I/O performance improvements, in particular offloading to OS and FS! :-)

File: test/jdk/java/nio/file/Files/CopyAndMove.java (hunk 1 above)
Comment (bplb) [reply test/jdk/java/nio/file/Files/CopyAndMove.java:52]: I agree about having OS-specific `CopyFile` classes instead of OS-specific providers. I don’t like the `IOStatus` being in the provider and the call of `provider.clone()` seems odd.

- `LinuxNativeDispatcher.ioctl_ficlone` uses `Blocker` emulating the calls in `UnixNativeDispatcher`.
- `FICLONE` is the descendent of `BTRFS_IOC_CLONE` and has the same numerical value so I think it is fairly safe.

Agreed about the other points.

**Comment by bplb** [comment]
I agree about having OS-specific `CopyFile` classes instead of OS-specific providers. I don’t like the `IOStatus` being in the provider and the call of `provider.clone()` seems odd.

- `LinuxNativeDispatcher.ioctl_ficlone` uses `Blocker` emulating the calls in `UnixNativeDispatcher`.
- `FICLONE` is the descendent of `BTRFS_IOC_CLONE` and has the same numerical value so I think it is fairly safe.

Agreed about the other points.

File: test/jdk/java/nio/file/Files/CopyAndMove.java (hunk 1 above)
Comment (bplb) [reply test/jdk/java/nio/file/Files/CopyAndMove.java:52]: Most testing is done on ext4 so this is covered.

File: src/java.base/aix/classes/sun/nio/fs/CloneFile.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * Clones the file whose path name is {@code src} to that whose path
+     * name is {@code dst} using a platform-specific system call.
+     *
+     * @implSpec
+     * The implementation in this class always returns
+     * {@code IOStatus.UNSUPPORTED}.
+     *
+     * @param src the path of the source file
+     * @param dst the path of the desintation file (clone)
```
Comment (turbanoff) [inline src/java.base/aix/classes/sun/nio/fs/CloneFile.java:46]: typo `desintation`. It seems it was copied in all CloneFile.java files.

Comment (bplb) [reply src/java.base/aix/classes/sun/nio/fs/CloneFile.java:46]: > typo `desintation`. It seems it was copied in all CloneFile.java files.

Thanks, I think I need a spell checker.

Comment (bplb) [reply src/java.base/aix/classes/sun/nio/fs/CloneFile.java]: Fixed in 9810352d8ba77fd0de56ac24540c3714400a455e.

**Comment by AlanBateman** [comment]
I skimmed the latest patch and I think the main thing that still needs to be worked out is how to integtrate "clonefile". Right now the current prototype adds sun.nio.fs.CloneFile but that doesn't fit into the provider architecture. I think we are getting close to the point with these changes where UnixNativeDispatcher will need to be changed to define instance methods rather than static methods. I'll add comments soon with suggestions for how this could be integrated.

**Comment by bplb** [comment]
> I skimmed the latest patch and I think the main thing that still needs to be worked out is how to integtrate "clonefile". Right now the current prototype adds sun.nio.fs.CloneFile but that doesn't fit into the provider architecture. [...]

I don't like this approach that much either but is better than two prior versions I did not publish which respectively used a singleton version of `UnixCopyFile` and a `FunctionalInterface` for `clonefile`.

**Comment by AlanBateman** [comment]
Thanks for the update. One thing that I should have pointed out in previous comments is UnixFileSystem.copyNonPosixAttributes. This is overridden in LinuxFileSystem and others and I think gives you an example of how to add the clone method without needing fileCopier. Would you be able to try that? I suspect the patch will reduce if that is used as the template.

**Comment by bplb** [comment]
> [...] I think gives you an example of how to add the clone method without needing fileCopier. Would you be able to try that? I suspect the patch will reduce if that is used as the template.

I'll take a look. If it improves things, then so much the better.

**Comment by bplb** [comment]
`clone()` moved to `{Bsd,Linux,Unix}FileSystem`; `{Bsd,Linux}CopyFile` removed.

File: src/java.base/macosx/native/libnio/fs/BsdNativeDispatcher.c [hunk 3]
```diff
@@ -261,7 +279,8 @@ Java_sun_nio_fs_BsdNativeDispatcher_setattrlist0(JNIEnv* env, jclass this,
     attrList.bitmapcount = ATTR_BIT_MAP_COUNT;
     attrList.commonattr = commonattr;
 
-    if (setattrlist(path, &attrList, (void*)buf, count*attrsize, options) != 0) {
+    if (setattrlist(path, &attrList, (void*)buf,
+        count*attrsize, options) != 0) {
```
Comment (AlanBateman) [inline src/java.base/macosx/native/libnio/fs/BsdNativeDispatcher.c:283]: I assume this is an accidental edit.

File: src/java.base/macosx/native/libnio/fs/BsdNativeDispatcher.c [hunk 4]
```diff
@@ -225,6 +227,22 @@ Java_sun_nio_fs_BsdNativeDispatcher_getmntonname0(JNIEnv *env, jclass this,
     return mntonname;
 }
 
+JNIEXPORT jint JNICALL
+Java_sun_nio_fs_BsdNativeDispatcher_clonefile0(JNIEnv* env, jclass this,
+    jlong srcAddress, jlong dstAddress, int flags)
```
Comment (AlanBateman) [inline src/java.base/macosx/native/libnio/fs/BsdNativeDispatcher.c:232]: flags should be a "jint" (not int).

File: src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java [hunk 5]
```diff
@@ -64,6 +64,29 @@ static byte[] getmntonname(UnixPath path) throws UnixException {
     }
     static native byte[] getmntonname0(long pathAddress) throws UnixException;
 
+    /**
+     * int clonefile(const char * src, const char * dst, int flags);
+     */
+    static int clonefile(UnixPath src, UnixPath dst, int flags)
+        throws UnixException {
+        if (src.getFileSystem() == dst.getFileSystem()) {
```
Comment (AlanBateman) [inline src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java:72]: Why does this check if getFileSystem() returns the same file system? There is only one instance of UnixFileSystem so this will always be true.

Will the "throws UnixException {" fit on the previous line to fixing the odd formatting.

File: src/java.base/macosx/classes/sun/nio/fs/BsdFileSystem.java [hunk 6]
```diff
@@ -67,6 +70,38 @@ public Set<String> supportedFileAttributeViews() {
...
+     * @param followLinks whether to follow links
+     *
+     * @return 0 on success, IOStatus.UNSUPPORTED_CASE if the call does not work
+     *         with the given parameters, or IOStatus.UNSUPPORTED if cloning is
+     *         not supported on this platform
+     */
+    @Override
+    protected int clone(UnixPath src, UnixPath dst, boolean followLinks)
+        throws IOException {
+        int flags = followLinks ? 0 : CLONE_NOFOLLOW;
```
Comment (AlanBateman) [inline src/java.base/macosx/classes/sun/nio/fs/BsdFileSystem.java:88]: Can you put the "{" onto its own line so that it's a bit easier to distinguish the method declaration from the body.

File: src/java.base/linux/native/libnio/fs/LinuxNativeDispatcher.c [hunk 7]
```diff
@@ -36,6 +38,10 @@
 
 #include "sun_nio_fs_LinuxNativeDispatcher.h"
 
+#ifndef FICLONE
+#define FICLONE      1074041865
```
Comment (AlanBateman) [inline src/java.base/linux/native/libnio/fs/LinuxNativeDispatcher.c:42]: Are we sure this value is the same on all architectures?

File: src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java [hunk 8]
```diff
@@ -67,6 +71,115 @@ public Set<String> supportedFileAttributeViews() {
...
+        try {
+            srcFD = open(src, O_RDONLY, 0);
+        } catch (UnixException x) {
+            x.rethrowAsIOException(src);
+            return IOStatus.THROWN;
+        }
+
+        int dstFD = 0;
+        try {
+            dstFD = open(dst, O_CREAT | O_WRONLY, 0666);
```
Comment (AlanBateman) [inline src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java:111]: The mode should be attrs.mode() rather than 0666, which means the clone method need an additional parameter.

File: src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java [hunk 9]
```diff
@@ -67,6 +71,115 @@ public Set<String> supportedFileAttributeViews() {
...
+     * @param src the path of the source file
+     * @param dst the path of the destination file (clone)
+     * @param followLinks whether to follow links
+     *
+     * @return 0 on success, IOStatus.UNSUPPORTED_CASE if the call does not work
+     *         with the given parameters, or IOStatus.UNSUPPORTED if cloning is
+     *         not supported on this platform
+     */
+    protected int clone(UnixPath src, UnixPath dst, boolean followLinks)
+        throws IOException {
```
Comment (AlanBateman) [inline src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java:100]: Formatting nit here too.

File: src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java [hunk 10]
```diff
@@ -67,6 +71,115 @@ public Set<String> supportedFileAttributeViews() {
         return SupportedFileFileAttributeViewsHolder.supportedFileAttributeViews;
     }
 
+    private static UnixException catEx(UnixException x, UnixException y) {
```
Comment (AlanBateman) [inline src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java:74]: What is the purpose of add a suppressed exception to a UnixException? A UnixException will get re-mapped to a File SystemException so I assume the suppressed exception will be ignored.

File: src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java [hunk 11]
```diff
@@ -273,6 +276,18 @@ private static void copyFile(UnixPath source,
                                  long addressToPollForCancel)
         throws IOException
     {
+        boolean copied = false;
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java:279]: The "copied" flag is a sign that the copyFile method needs to be refactor to split it into the copy of the contents and the file attributes. Also on first glance, re-opening both the source and destination files after a clone is a bug. We should (where possible) only open the source file for reading and the destination file for writing once.

Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java:279]: On macOS it's not possible to clone using file descriptors; paths are required.

Comment (AlanBateman) [reply src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java:279]: > On macOS it's not possible to clone using file descriptors; paths are required.

I understand that but what isn't clear is if the attributes are cloned too.  It might be that we only use it when not copying attributes.

File: src/java.base/linux/native/libnio/fs/LinuxNativeDispatcher.c (hunk 7 above)
Comment (bplb) [reply src/java.base/linux/native/libnio/fs/LinuxNativeDispatcher.c:42]: Testing succeeded on all platforms I've tried. It's been around since [Linux 4.5](https://man7.org/linux/man-pages/man2/ioctl_ficlone.2.html#VERSIONS) which was released [13 March 2016](https://en.wikipedia.org/wiki/Linux_kernel_version_history#Releases_4.x.y).

File: src/java.base/macosx/native/libnio/fs/BsdNativeDispatcher.c (hunk 3 above)
Comment (bplb) [reply src/java.base/macosx/native/libnio/fs/BsdNativeDispatcher.c]: Right; fixed in 2157a1480229fa697c824ae814f11a309627d768.

File: src/java.base/macosx/native/libnio/fs/BsdNativeDispatcher.c (hunk 4 above)
Comment (bplb) [reply src/java.base/macosx/native/libnio/fs/BsdNativeDispatcher.c]: Fixed in 2157a1480229fa697c824ae814f11a309627d768.

File: src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java (hunk 5 above)
Comment (bplb) [reply src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java]: Fixed in 2157a1480229fa697c824ae814f11a309627d768.

File: src/java.base/macosx/classes/sun/nio/fs/BsdFileSystem.java (hunk 6 above)
Comment (bplb) [reply src/java.base/macosx/classes/sun/nio/fs/BsdFileSystem.java:91]: Fixed in 2157a1480229fa697c824ae814f11a309627d768.

File: src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java (hunk 8 above)
Comment (bplb) [reply src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java]: Fixed in 2157a1480229fa697c824ae814f11a309627d768.

File: src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java (hunk 9 above)
Comment (bplb) [reply src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java]: Fixed in 2157a1480229fa697c824ae814f11a309627d768.

File: src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java (hunk 10 above)
Comment (bplb) [reply src/java.base/linux/classes/sun/nio/fs/LinuxFileSystem.java]: It was a mistake; fixed in 2157a1480229fa697c824ae814f11a309627d768.

File: src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java (hunk 11 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java]: Based on my testing, attributes appear to be cloned by default except for the permission on linux-aarch64 and the (group) ownership on macosx-x64. If we want to suppress that behavior, then using clone would have to be made conditional.

**Comment by AlanBateman** [comment]
I took another pass over the current patch. It has improved significantly but I think we will need to restructure UnixCopyFile very significantly to allow for both clone implementations. The main issue is that the macOS clonefile is forcing the patch to clone with file paths and this breaks architecture where the source and target files are opened once.

One thing we could do is separate the update to FileChannelImpl.c to use copy_file_range to a separate issue. It doesn't need to be in this PR.

Another thing we could do is separate out the use of ioctl(FICLONE) on Linux. That is essentially an alternative transfer implementation so if we move the transfer to UnixFileSystem then the Linux implementation can use ioctl(FICLONE) or sendfile64, the macOS implementation can use fcopyfile, and the "portable" Unix implementation can use the user-space transfer loop. The implementation specific can fallback to the Unix transfer if needed.  If we do this then it gives us more time to think about the troublesome macOS clonefile.

**Comment by AlanBateman** [comment]
I've skimmed through ecce2093. There is a lot of refactoring (which is good) and it also introduces the use of clone on macOS. What would you think about separating this two? Let's get the Linux and BSD/macOS separation out of the way and that would allow us to discuss the semantics issue and agree when it is safe to use clone.

**Comment by bplb** [comment]
> What would you think about separating this two?

That would be fine with me, but I think the BSD / Linux separation should be done under a different issue with the `clonefile` work remaining under this one. The addition of `copy_file_range` function is also more pertinent to this issue than to refactoring however so I don't know under which that should go.

**Comment by bplb** [comment]
> > That would be fine with me, but I think the BSD / Linux separation should be done under a different issue with the `clonefile` work remaining under this one.
> 
> That's okay, I think the main important point is that the refactoring is worthy on its own [...].

Probably then I'll file two new issues: one for the refactoring and a sub-task for `copy_file_range` and change the topic of the current sub-task to `clonefile`. The sub-tasks will both be blocked by the refactoring.

File: src/java.base/linux/native/libnio/fs/LinuxFileSystem.c [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+    if (my_copy_file_range_func != NULL) {
+        do {
+            RESTARTABLE(my_copy_file_range_func(src, NULL, dst, NULL, count, 0),
+                                                bytes_sent);
+            if (bytes_sent < 0) {
+                switch (errno) {
+                    case EINVAL:
+                    case EXDEV:
+                        // ignore and try sendfile()
```
Comment (bplb) [inline src/java.base/linux/native/libnio/fs/LinuxFileSystem.c:96]: `ENOSYS` will need to be handled as done [here](https://github.com/openjdk/jdk/pull/10055/files#diff-4fdc393c9569db35d35d0b5698849aa939ba395ec5a685befadbb2bacbaddb3e).

**Comment by AlanBateman** [comment]
The updated implementation changes in 62461a2d look okay but I'm not sure about the updates to the CopyAndMove test. If I read it correctly, it tries to find the mount point where a COW file system is mounted and use it as an additional test directory. I don't think we should want tests trying to use these locations as test directories, even if running with permissions.

**Review by AlanBateman — APPROVED** [review]

**Comment by mikehearn** [comment]
Thank you! I'm the one who filed the original webbug report (some day I must figure out how to get direct access to the JDK bug tracker). 

This wasn't a theoretical request, it was motivated by a real product that my company has just launched called [Conveyor](https://conveyor.hydraulic.dev/). Conveyor is a sort of jpackage++ and is structured as a parallel build system in which intermediate steps are cached. The different tasks sometimes have to copy large files around (like JDKs, JARs etc) and using kernel level COW cloning will yield a large boost to performance on filesystems that support it.
